ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp

```diff
@@ -1107,6 +1111,16 @@ void Matmul::validate(
         a_shape[-1],
         b_shape[-2]);
 
+    const bool is_optional_output_tensor = !optional_output_tensors.empty() && optional_output_tensors.at(0).has_value();
+    if (is_optional_output_tensor) {
+        const auto& optional_output_tensor_c = optional_output_tensors.at(0);
+        const auto& optional_output_tensor_shape = optional_output_tensor_c->get_logical_shape();
+        const auto output_tensor_spec = this->compute_output_specs(input_tensors, {}).at(0);
+        TT_FATAL(
+            optional_output_tensor_shape == output_tensor_spec.logical_shape(),
+            "Shape of Optional Output Tensor should match Output Tensor");
```

Comment:
We need to also validate for output dtype and memory config (especially for sharding in memory config). This is because we use the passed in or default memory config of the matmul struct to select which matmul program to run. (See how `generate_matmul_program_config` is called in `get_program_config`.). So we need to either:
- Assert out if provided specs through output dtype, memory config don't match the specs of the optional output tensor
- Override (and maybe throw a warning) the passed in args so the stored memory config matches the optional output tensor's

Reply:
We should revisit. Option 1 is not really a good solution. Imagine if some user just called matmul with two inputs and one optional output where the optional output specs don't match with the defaults set by matmul. It will always error out in that case.

Reply:
`compute_output_specs()` always returns `optional_output_tensor`'s specs if its not empty
Therefore, this assert condition will be true

Reply:
@TT-BrianLiu this is intended to be used by automated tools to avoid memory fragmentation. Given that, I think it should be okay to make the requirements stringent. Would you be okay with that if the doc string has a warning that the requirements are strict?

Reply:
I talked to @TT-BrianLiu 

If optional output tensor is specified, then dtype and memory config need to be checked.
- if they are default then they should be set based on optional output tensor
- if the are not default then they should be compared and if there is a difference an error reported (via fatal or throw)

The doc string for optional output tensor should convey this information.

output_dtype is an optional and in ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp create_matmul_struct() we do
parameters.output_dtype.value_or(input_tensor_a.get_dtype())
we need to expand the code to have extra logic.

For parameters.output_mem_config we use DEFAULT_OUTPUT_MEMORY_CONFIG as a default (if the user specifies that or None). We'd need to just overwrite if we see that if the optional output tensor is specified.

Checks should be done in create_matmul_struct() as part of the extra logic.

We should retain these checks below in validate as well in case someone decides to create a separate code path into matmul.

Reply:
@bbradelTT this is addressed in latest patch